language/resource-api/README.md

```diff
@@ -333,33 +344,166 @@ When a resource is marked with `noop => true`, either locally or through a globa
 ### Provider feature: `remote_resource`
 
 ```ruby
+# lib/puppet/type/nx9k_vlan.rb
 Puppet::ResourceApi.register_type(
   name: 'nx9k_vlan',
   features: [ 'remote_resource' ],
+  # ...
 )
 
-require 'puppet/util/network_device/simple/device'
-module Puppet::Util::NetworkDevice::Nexus
-  class Device < Puppet::Util::NetworkDevice::Simple::Device
-    def facts
-      # access the device and return facts hash
-    end
-  end
-end
-
-class Puppet::Provider::Nx9k_vlan::Nx9k_vlan
+# lib/puppet/provider/nx9k_vlan/nexus.rb
+class Puppet::Provider::Nx9k_vlan::Nexus
   def set(context, changes, noop: false)
     changes.each do |name, change|
       is = change.has_key? :is ? change[:is] : get_single(name)
       should = change[:should]
-      # ...
-      context.device.do_something unless noop
+
+      established = context.transport.verify
+      if established.nil?
+        context.transport.do_something unless noop
+      else
+        raise Puppet::Transport::Error('Connection to remote target is not available: #{established}')
+      end
+    end
+  end
+```
+
+Declaring this feature restricts the resource from being run "locally". It executes all external interactions through the `context.transport` instance. The way the instance is set up is runtime specific. In Puppet, it is configured through the [`device.conf`](https://puppet.com/docs/puppet/5.3/config_file_device.html) file, and only available when running under [`puppet device`](https://puppet.com/docs/puppet/5.3/man/device.html).
+
+To support Puppet versions prior to 6, see the [Legacy Support](#legacy-support) section below.
+
+### Transport
+
+```ruby
+# lib/puppet/transport/nexus_schema.rb
+Puppet::ResourceAPI.register_transport(
+  name: 'nexus', # points at class Puppet::Transport::Nexus
+  desc: 'Connects to a Cisco Nexus device'
+  features: [], # TODO
+  connection_info: {
+    connector_type: {
+      type: 'Enum[nxapi, ssh, grpc, pidgeon]',
```

Comment:
Will this be supporting the `Optional` data type?

Reply:
The Resource API does not put any restrictions on the types folks can use here.

That said, I would understand if you want to add guidance to developers around what can be mapped into UI (String (including Pattern and Enum), Number, Variant, Optional, Array, sensitive) and do a generic JSON(?) input field for everything else.